lib/graphql/language/nodes.rb

```diff
@@ -25,14 +25,35 @@ def initialize_node(options={})
           raise NotImplementedError
         end
 
+        def eql?(other)
+          other.is_a?(self.class) &&
+            other.scalars.eql?(self.scalars) &&
+            other.children.eql?(self.children)
+        end
+
         # @return [GraphQL::Language::Nodes::AbstractNode] all nodes in the tree below this one
         def children
           self.class.child_attributes
             .map { |attr_name| public_send(attr_name) }
             .flatten
         end
 
+        def scalars
+          self.class.scalar_attributes
+            .map { |attr_name| public_send(attr_name) }
+        end
+
         class << self
+          def inherited(subclass)
+            subclass.scalar_attributes(*@scalar_attributes)
+            subclass.child_attributes(*@child_attributes)
+          end
```

Comment:
This was an existing bug in `child_attributes` as well. If a subtype extends another class like `class ListType < WrapperType; end`, it won't inherit class ivars. You always need to copy them. But that ensures they get their own copy of the Array. Which is good.